File: go/lang/wrapperoopiterable.go
package lang

type wrapperOOPIterable[FROM any, TO any] struct {
	slice []FROM
	wrap func(FROM) TO
	to TO
}

func (this wrapperOOPIterable[FROM,TO]) OOPIterator() OOPIterator[TO] {
	return WrapperOOPIteratorWrapSlice(this.slice, this.wrap, this.to)
}


func WrapperOOPIterableWrapSlice[OOP any, PURE any](slice []PURE, wrap func(PURE) OOP, oopClass OOP) OOPIterable[OOP] {
	ret := wrapperOOPIterable[PURE, OOP] {
		slice,
		wrap,
		oopClass,
	}
	return ret
}

func WrapperOOPIterableUnwrapSlice[OOP any, PURE any](slice OOPIterable[OOP], unwrap func(OOP) PURE, pureClass PURE) []PURE {
	return WrapperOOPIteratorUnwrapSlice(slice.OOPIterator(), unwrap, pureClass)
}

File: python/lang/otterop/lang/wrapper_oop_iterable.py
from otterop.lang.wrapper_oop_iterator import WrapperOOPIterator


class WrapperOOPIterable:

    def __init__(self, iterable, wrap):
        self._iterable = iterable
        self._wrap = wrap or (lambda x: x)

    def oop_iterator(self):
        return WrapperOOPIterator(self._iterable.__iter__(), self._wrap)

    def __iter__(self):
        return WrapperOOPIterator(self._iterable.__iter__(), self._wrap)

    @staticmethod
    def wrap(iterable, wrap):
        if wrap is None and isinstance(iterable, WrapperOOPIterable):
            return iterable
        return WrapperOOPIterable(iterable, wrap)

    @staticmethod
    def unwrap(oop_iterable, unwrap):
        if unwrap is None:
            return oop_iterable
        return WrapperOOPIterable(oop_iterable, unwrap)
